tfx/components/base/base_executor.py

```diff
@@ -160,3 +160,119 @@ def Do(self, input_dict: Dict[Text, List[types.Artifact]],
+                  output_dict: Dict[Text, List[types.Artifact]],
+                  exec_properties: Dict[Text, Union[int, float, Text]]
+                  ) -> Dict[Text, beam.pvalue.PCollection]:
+    """Executes reads for underlying component implementation.
+
+    Args:
+      input_dict: Input dict from input key to a list of Artifacts. These are
+        often outputs of another component in the pipeline and passed to the
+        component by the orchestration system.
+      output_dict: Output dict from output key to a list of Artifacts. These are
+        often consumed by a dependent component.
+      exec_properties: A dict of execution properties. These are inputs to
+        pipeline with primitive types (int, string, float) and fully
+        materialized when a pipeline is constructed. No dependency to other
+        components or later injection from orchestration systems is necessary or
+        possible on these values.
+
+    Returns:
+      A dict from input artifact name/split key to PCollections from reads for
+      underlying component execution logic.
+    """
+    pass
+
+  @abc.abstractmethod
+  def run_component(self, pipeline: beam.Pipeline,
+                    beam_inputs: Dict[Text, beam.pvalue.PCollection],
+                    input_dict: Dict[Text, List[types.Artifact]],
+                    output_dict: Dict[Text, List[types.Artifact]],
+                    exec_properties: Dict[Text, Union[int, float, Text]]
+                    ) -> Dict[Text, beam.pvalue.PCollection]:
+    """Executes underlying component implementation, except for reads/writes.
+
+    Args:
+      beam_inputs: A Beam input dict from input key to PCollections. Output
+        of self.read_inputs().
+      input_dict: Input dict from input key to a list of Artifacts. These are
+        often outputs of another component in the pipeline and passed to the
+        component by the orchestration system.
+      output_dict: Output dict from output key to a list of Artifacts. These are
+        often consumed by a dependent component.
+      exec_properties: A dict of execution properties. These are inputs to
+        pipeline with primitive types (int, string, float) and fully
+        materialized when a pipeline is constructed. No dependency to other
+        components or later injection from orchestration systems is necessary or
+        possible on these values.
+
+    Returns:
+      A dict from output artifact name/split key to PCollections from underlying
+      component execution logic (not including reads/writes).
+    """
+    pass
+
+  @abc.abstractmethod
+  def write_outputs(self, pipeline: beam.Pipeline,
+                    beam_outputs: Dict[Text, beam.pvalue.PCollection],
+                    input_dict: Dict[Text, List[types.Artifact]],
+                    output_dict: Dict[Text, List[types.Artifact]],
+                    exec_properties: Dict[Text, Union[int, float, Text]]
+                    ) -> None:
```

Comment:
Question for @charlesccychen: In Beam Python, sink PTransforms typically return PCollections that represent when the write has been completed; should those PCollections be exposed in the return value here?

Don't need to change this in this PR; just a thought.